WalletWasabi/WabiSabi/Backend/Rounds/Arena.cs

```diff
@@ -391,39 +395,68 @@ public async Task<ConnectionConfirmationResponse> ConfirmConnectionAsync(Connect
+			var vsizeZeroTask = Task.Run(() => round.VsizeCredentialIssuer.PrepareResponse(request.ZeroVsizeCredentialRequests), cancellationToken);
+			Task<ICommitableCredentialsResponse>? amountRealTask = null;
+			Task<ICommitableCredentialsResponse>? vsizeRealTask = null;
 
-					// update state
-					alice.ConfirmedConnection = true;
-					round.CoinjoinState = state;
+			if (round.Phase is Phase.ConnectionConfirmation)
+			{
+				amountRealTask = Task.Run(() => round.AmountCredentialIssuer.PrepareResponse(realAmountCredentialRequests), cancellationToken);
+				vsizeRealTask = Task.Run(() => round.VsizeCredentialIssuer.PrepareResponse(realVsizeCredentialRequests), cancellationToken);
+			}
 
-					return new(
-						commitAmountZeroCredentialResponse.Commit(),
-						commitVsizeZeroCredentialResponse.Commit(),
-						commitAmountRealCredentialResponse.Commit(),
-						commitVsizeRealCredentialResponse.Commit());
+			using (await AsyncLock.LockAsync(cancellationToken).ConfigureAwait(false))
+			{
+				if (round.Alices.FirstOrDefault(x => x.Id == request.AliceId) != alice)
+				{
+					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.AliceNotFound, $"Round ({request.RoundId}): Alice ({request.AliceId}) not found.");
 				}
-				else
+
+				switch (round.Phase)
 				{
-					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.WrongPhase, $"Round ({request.RoundId}): Wrong phase ({round.Phase}).");
+					case Phase.InputRegistration:
+						{
+							var commitAmountZeroCredentialResponse = await amountZeroTask.ConfigureAwait(false);
+							var commitVsizeZeroCredentialResponse = await vsizeZeroTask.ConfigureAwait(false);
+							alice.SetDeadlineRelativeTo(round.ConnectionConfirmationTimeout);
+							return new(
+								commitAmountZeroCredentialResponse.Commit(),
+								commitVsizeZeroCredentialResponse.Commit());
+						}
+
+					case Phase.ConnectionConfirmation:
+						{
+							// If the phase was InputRegistration before then we did not pre-calculate real credentials.
+							amountRealTask ??= Task.Run(() => round.AmountCredentialIssuer.PrepareResponse(realAmountCredentialRequests), cancellationToken);
+							vsizeRealTask ??= Task.Run(() => round.VsizeCredentialIssuer.PrepareResponse(realVsizeCredentialRequests), cancellationToken);
+
+							var commitAmountZeroCredentialResponse = await amountZeroTask.ConfigureAwait(false);
+							var commitVsizeZeroCredentialResponse = await vsizeZeroTask.ConfigureAwait(false);
+							var commitAmountRealCredentialResponse = await amountRealTask.ConfigureAwait(false);
+							var commitVsizeRealCredentialResponse = await vsizeRealTask.ConfigureAwait(false);
+
+							// Ensure the input can be added to the CoinJoin
+							round.CoinjoinState = round.Assert<ConstructionState>().AddInput(alice.Coin);
+
+							// update state
+							alice.ConfirmedConnection = true;
```

Comment:
This should be set to `true` only after we have the `ConnectionConfirmationResponse` because otherwise if we get an error there the Alice will be in an invalid state.

Reply:
What about `round.CoinjoinState = round.Assert<ConstructionState>().AddInput(alice.Coin);`? Is that in the right place?

Reply:
the current behavior is correct, `AddInput` is the last thing that can error  `Commit` will succeed unconditionally and is in fact redundant now that we changed `CredentialIssuer`, so I already have a local branch that removes it that I will PR after these are merged to keep the noise down, so afterb that is merged this will be the last statement here.